soc/silabs/silabs_s2/soc.c

```diff
@@ -39,3 +52,53 @@ void soc_early_init_hook(void)
 	sl_hfxo_manager_init();
 #endif
 }
+
+#if defined(CONFIG_ARM_SECURE_FIRMWARE) && !defined(CONFIG_ARM_FIRMWARE_HAS_SECURE_ENTRY_FUNCS)
+static void smu_fault(void)
+{
+	PR_FAULT_INFO("***** SMU FAULT *****");
+
+	if (SMU->IF & SMU_IF_BMPUSEC) {
+		PR_FAULT_INFO("Bus Manager Fault");
+		PR_EXC("SMU.BMPUFS=%d", SMU->BMPUFS);
+	}
+	if (SMU->IF & SMU_IF_PPUSEC) {
+		PR_FAULT_INFO("Peripheral Access Fault");
+		PR_EXC("SMU.PPUFS=%d", SMU->PPUFS);
+	}
+
+	z_fatal_error(K_ERR_CPU_EXCEPTION, NULL);
+}
+#endif
+
+void soc_prep_hook(void)
+{
+	/* Initialize TrustZone state of the device.
+	 * If this is a secure app with no non-secure callable functions, it is a secure-only app.
+	 * Configure all peripherals except the SMU and SEMAILBOX to non-secure aliases, and make
+	 * all bus transactions from the CPU have non-secure attribution.
+	 * This makes the secure-only app behave more like a non-secure app, allowing the use of
+	 * libraries that only expect to use non-secure peripherals, such as the radio subsystem.
+	 */
+#if defined(CONFIG_ARM_SECURE_FIRMWARE) && !defined(CONFIG_ARM_FIRMWARE_HAS_SECURE_ENTRY_FUNCS)
```

Comment:
You can probably replace that by `if (IS_ENABLE(CONFIG_ARM_SECURE_FIRMWARE) && !IS_ENABLED(CONFIG_ARM_FIRMWARE_HAS_SECURE_ENTRY_FUNCS))`

Reply:
There are several reasons this won't work:
* `SAU` is only defined in `core_cm33.h` if `-mcmse` is passed to the compiler (which sets the `__ARM_FEATURE_CMSE` internal macro)
* Interrupt handler registering works by assigning a variable into a global section `.intList`, which causes the variable to "escape" the local dead code elimination by `if (0)`, and end up mapped in the vector table anyway.

So I think we'll have to keep using #if here.